### PR TITLE
Add minimal CQM implementation

### PR DESCRIPTION
**Pull request by arcondello**
Allows the construction of binary constrained quadratic models.

### CODE REVIEW

**Comment by arcondello** [comment]
Still need to fix the windows bug obviously. This also will need a pretty serious second pass when we add `QuadraticModel`

File: dimod/constrained.py [hunk 1]
```diff
@@ -0,0 +1,395 @@
...
+
+        data = dict(num_variables=len(self.variables),
+                    num_constraints=len(self.constraints),
+                    num_biases=self.num_biases(),
+                    )
+
+        header_data = json.dumps(data, sort_keys=True).encode('ascii')
+        header_data += b'\n'
+        header_data += b' '*(64 - (len(prefix) + 6 + len(header_data)) % 64)
+        header_len = np.dtype('<u4').type(len(header_data)).tobytes()
```
Comment (randomir) [inline dimod/constrained.py:364]: I did just a quick check, but it looks `.tobytes()` in this setup actually always uses **native** byteorder. I believe your intention was to use little endian.

**Review by randomir — APPROVED** [review]
Apart from a few minor issues, LGTM!

File: dimod/constrained.py [hunk 2]
```diff
@@ -0,0 +1,395 @@
...
+        return objective
+
+    def add_constraint(self, data, *args, **kwargs):
+        """A convenience wrapper for other methods that add constraints."""
+        if isinstance(data, BinaryQuadraticModel):
+            self.add_constraint_from_bqm(data, *args, **kwargs)
+        elif isinstance(data, Comparison):
+            self.add_constraint_from_comparison(data, *args, **kwargs)
+        elif isinstance(data, Iterable):
+            self.add_constraint_from_iterable(data, *args, **kwargs)
```
[the remark this answers is not in the document]
Comment (arcondello) [reply dimod/constrained.py:90]: Ah, used to have a comment about that, not sure where it went. But alas, `singledispatchmethod` is only available in 3.8+

[the remark this answers is not in the document]
Comment (randomir) [reply dimod/constrained.py:90]: Ah, I checked in docs for `singledispatch` (available from 3.4), but not for the method variant.

File: dimod/constrained.py (hunk 1 above)
Comment (arcondello) [reply dimod/constrained.py:364]: Yup, excellent catch
```
In [6]: np.frombuffer(np.dtype('<u4').type(256).tobytes(), '<u4')
Out[6]: array([256], dtype=uint32)

In [7]: np.frombuffer(np.dtype('>u4').type(256).tobytes(), '<u4')
Out[7]: array([256], dtype=uint32)

In [8]: np.frombuffer(np.dtype('>u4').type(256).tobytes(), '>u4')
Out[8]: array([65536], dtype=uint32)

In [9]: np.frombuffer(np.dtype('<u4').type(256).tobytes(), '>u4')
Out[9]: array([65536], dtype=uint32)
```

Comment (randomir) [reply dimod/constrained.py:365]: I see we have the same bug in BQM and DQM.

Comment (arcondello) [reply dimod/constrained.py:365]: Indeed. Investigating

Comment (arcondello) [reply dimod/constrained.py:365]: Have a fix I'll make in a followup PR.

**Pull request merged**